### DISCUSSION TITLE
Priority of back callbacks on children of components with +1 stacks?

### DISCUSSION THREAD

**Discussion by YektaDev** [arkivanov/Decompose#512] [Q&A] [answered]
Hi.

I wonder if the behavior of back callback invocation is calling by priority across children of a component with more than one stack? In my own little investigation I found different behavior on Android and desktop (It's too and abstracted away to provide a reproducer).
Having the following parent:
```kotlin
SomeComponent {
  > stack1
  > stack2
}
```
In desktop, It's possible to rely on callback priority for the children. Meaning if `stack2`'s active child registers a back callback later than other callbacks, it's invoked first. But in Android, as I investigated, it doesn't matter if the `stack2`'s active child is the last one to register a back callback, as `stack1`'s children are always prioritized in regards to registered callbacks _(still not sure if it's because of Android, Decompose, or my own code, I doubt it though)._

Thanks in advance!

_p.s. I feel handling back press becomes hard pretty quickly when it comes to more than one stack, especially recently._

**Comment by arkivanov** [comment, marked as answer]
There should be no difference in back handling depending on the platform, as all the code is multiplatform. Assuming you drive the root lifecycle on desktop (e.g. by using `LifecycleController`). Also it's not clear how do you trigger the back button on desktop?

The logic is pretty simple.

Each Child Stack registers two back callbacks when `childStack` function is called. One callback with priority equal to `PRIORITY_DEFAULT + 1` for the currently active child component, and another callback with priority equal to `PRIORITY_DEFAULT` for the stack itself. The former callback allows the active component to intercept the back button. This callbacks is only enabled when there is an inner enabled callback (e.g. a callback registered by the nested Child Stack or by a manually registered nested callback). The latter callbacks allows automatic stack popping, the callback is only enabled if the `handleBackButton` parameter is `true`.

So there is a tree of callbacks. When the back button is pressed, the call is propagated from the root down the tree. On each level the latest enabled callback with the highest priority is selected and the recursion continues until a leaf callback is found and called.

**Comment by YektaDev** [comment]
> Assuming you drive the root lifecycle on desktop (e.g. by using LifecycleController).

That's exactly the case.

> Also it's not clear how do you trigger the back button on desktop?

I hold a reference to the `BackDispatcher` passed to the root and invoke back press on `Esc` key, simply emulating a "Back Button" for desktop.

Thanks for the clarification! Now that the process is clear, I guess the issue is on me.

**Reply by arkivanov** [reply]
Oh, that's a very interesting application of BackDispatcher on Desktop! Thanks for sharing!
